Libraries/LibWeb/CSS/StyleComputer.cpp

```diff
@@ -1070,49 +1070,164 @@ void StyleComputer::collect_animation_into(DOM::AbstractElement abstract_element
 
-    auto start_composite_operation = to_composite_operation(keyframe_values.composite);
-    auto end_composite_operation = to_composite_operation(keyframe_end_values.composite);
-
-    for (auto const& it : computed_start_values) {
-        auto resolved_start_property = it.value;
-        RefPtr resolved_end_property = computed_end_values.get(it.key).value_or(nullptr);
-
-        if (!resolved_end_property) {
-            if (resolved_start_property) {
-                computed_properties.set_animated_property(it.key, *resolved_start_property);
-                dbgln_if(LIBWEB_CSS_ANIMATION_DEBUG, "No end property for property {}, using {}", string_from_property_id(it.key), resolved_start_property->to_string(SerializationMode::Normal));
+    // 5.3.4. The effect value of a keyframe effect
+    // https://drafts.csswg.org/web-animations-1/#the-effect-value-of-a-keyframe-animation-effect
+    struct KeyframeEffectInterpolateFailed { };
+    auto keyframe_effect = [&](PropertyID target_property) -> Variant<NonnullRefPtr<StyleValue const>, Empty, KeyframeEffectInterpolateFailed> {
+        // 1. If iteration progress is unresolved abort this procedure.
+        // NB: handled at the start
+        auto iteration_progress = output_progress.value();
+        // 2. Let target property be the longhand property for which the effect value is to be calculated
+        // NB: function parameter
+        // 3. If animation type of the target property is not animatable abort this procedure since the effect cannot be applied.
+        if (!CSS::is_animatable_property(target_property)) {
+            return Empty {};
+        }
+        // 4. If the keyframe effect does not have an effect target, or if the effect target cannot have computed property
+        //    values calculated, abort this procedure.
+        if (effect->target() == nullptr) {
+            return Empty {};
+        }
+
+        // 5. Define the neutral value for composition as a value which, when combined with an underlying value using the add
+        //    composite operation, produces the underlying value.
+        // AD-HOC: To reduce complexity we'll use composite operation of REPLACE in conjunction with the underlying value
+        //         https://github.com/w3c/csswg-drafts/issues/2914
+        NonnullRefPtr<StyleValue const> const& underlying_value = computed_properties.property(target_property);
+        // 6. Let property-specific keyframes be the result of getting the set of computed keyframes for this keyframe effect.
+        Vector<Animations::KeyframeEffect::KeyFrameSet::ComputedKeyframe> property_specific_keyframes;
+        // 7. Remove any keyframes from property-specific keyframes that do not have a property value for target property.
+        for (auto& keyframe : computed_keyframes) {
+            if (keyframe.properties.contains(target_property)) {
+                property_specific_keyframes.append(keyframe);
             }
-            continue;
         }
```

Comment:
Maybe put these spec steps together, with a note saying we're adding the ones we want instead of removing the ones we don't.